kernel: Expose `CheckTransaction` consensus validation function

w0xlt opened the pull request:
This PR exposes the consensus-level `CheckTransaction` function through the libbitcoinkernel C API and adds a corresponding C++ wrapper.

Currently, libkernel only provided script-level validation via `btck_script_pubkey_verify` and `ScriptPubkeyApi<>::Verify`. 

AFAIK there was no way to perform context-free consensus checks on a transaction’s structure (e.g., coinbase rules, money-range, duplicate inputs).

This change introduces a new API:

```c
int btck_check_transaction(const btck_Transaction* tx, btck_TxValidationState** out_state);
```

and a C++ convenience wrapper:

```cpp
std::pair<bool, TxValidationState> btck::CheckTransaction(const Transaction& tx);
```

Both follow the ownership and error-handling conventions established in `bitcoinkernel.h`.

The test suite is extended with cases covering:
* coinbase scriptSig length bounds
* empty vin / vout detection
* negative or out-of-range output values
* duplicate inputs
* null prevouts in non-coinbase transactions

w0xlt committed:
kernel: Expose btck_transaction_check consensus function

Add btck_transaction_check() to the libbitcoinkernel C API, exposing
context-free transaction consensus validation (consensus/tx_check.h).

Introduces btck_TxValidationState with introspection and lifecycle
functions. btck_TxValidationResult is exposed for compatibility with
existing validation-state APIs, though btck_transaction_check currently
reaches only UNSET and CONSENSUS.

Includes C++ wrapper and test coverage for btck_transaction_check using
test vectors from tx_valid.json / tx_invalid.json.

yuvicc commented on src/kernel/bitcoinkernel.h:
```diff
@@ -560,6 +593,19 @@ BITCOINKERNEL_API void btck_transaction_destroy(btck_Transaction* transaction);
+/**
+ * @brief Run consensus/tx_check::CheckTransaction on a tx and return the filled state.
+ * @return 1 if valid, 0 if invalid.
+ */
```
```suggestion
/**
   * @brief Run consensus/tx_check::CheckTransaction on a transaction.
   * 
   * Performs context-free consensus validation on a transaction without
   * requiring blockchain state.
   * 
   * @param[in] tx The transaction to validate
   * @param[out] out_state Pointer to receive the validation state (always set,
   *                       caller must destroy with btck_tx_validation_state_destroy)
   * @return 1 if valid, 0 if invalid
   */
```

w0xlt replied:
Done. Thanks.

yuvicc reviewed (commented):
Approach ACK

Great addition exposing `CheckTransaction` to the kernel API. This will be useful to external users needing context-free transaction validation.

janb84 reviewed (commented):
Concept ACK a262282abefc0a16be329ab2179c193b4b32cc5b

PR extends bitcoin kernel library with the checkTransaction consensus validation function.

Added a suggestion NIT (non-blocking) and I agree with the NIT from [yuvicc](https://github.com/bitcoin/bitcoin/pull/33796/files#r2506321358)

Steps taken: 
- Build & tested
- Code review
- Create a implementation PR for [BitcoinKernel.Net](https://github.com/janb84/BitcoinKernel.NET/pull/2)

w0xlt commented on src/test/kernel/test_kernel.cpp:
```diff
@@ -973,3 +1069,133 @@ BOOST_AUTO_TEST_CASE(btck_chainman_regtest_tests)
+
+// -----------------------------------------------------------------------------
+// CheckTransaction tests
+// -----------------------------------------------------------------------------
```
Done. Thanks.

w0xlt commented:
@sedited tests simplified.  Vectors from test/data/tx_invalid.json are now being reused.

sedited commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -9,6 +9,7 @@
 #include <chain.h>
 #include <coins.h>
 #include <consensus/validation.h>
+#include <consensus/tx_check.h>
```
Nit (ordering): This should be one line higher.

w0xlt replied:
Done. Thanks.

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -492,6 +514,17 @@ typedef uint8_t btck_ChainType;
+BITCOINKERNEL_API btck_TxValidationResult btck_tx_validation_state_get_tx_validation_result(
+    const btck_TxValidationState* state) BITCOINKERNEL_ARG_NONNULL(1);
+
+BITCOINKERNEL_API void btck_tx_validation_state_destroy(btck_TxValidationState* state);
```
Done. Thanks.

w0xlt commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1394,3 +1396,46 @@ void btck_block_header_destroy(btck_BlockHeader* header)
+btck_ValidationMode btck_tx_validation_state_get_validation_mode(const btck_TxValidationState* state_)
+{
+    const auto& state = btck_TxValidationState::get(state_);
+    if (state.IsValid())   return btck_ValidationMode_VALID;
```
Done. Thanks.

sedited reviewed (commented):
Thanks for applying the fixes, found something else that should be addressed.

w0xlt commented:
@sedited  Addressed the three actionable points in f490d9ec3d1a8d3f45d18b7ae6b9f85867fedc78.

Renamed the C entry point to `btck_transaction_check` to match the existing `namespace_noun_verb` convention. I also changed the API to take a caller-owned `btck_TxValidationState*` and added `btck_tx_validation_state_create()`, so callers can safely reuse an existing state object and we no longer allocate through `**` in the check function. . The C++ wrapper and kernel tests were updated accordingly, including a regression test that reuses the same validation state across multiple calls.

For the enum exposure comment, I did not change behavior `btck_transaction_check` still only reaches `UNSET` and `CONSENSUS` today, but I documented that explicitly while keeping the full enum exposed for forward compatibility.

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -625,6 +675,26 @@ BITCOINKERNEL_API void btck_precomputed_transaction_data_destroy(btck_Precompute
+ *                       The caller must destroy it with btck_tx_validation_state_destroy.
+ * @return               1 if valid, 0 if invalid.
+ */
+BITCOINKERNEL_API int BITCOINKERNEL_WARN_UNUSED_RESULT btck_check_transaction(
```
Thanks. Done in  https://github.com/bitcoin/bitcoin/commit/f490d9ec3d1a8d3f45d18b7ae6b9f85867fedc78.

alexanderwiederin reviewed (changes requested):
Concept ACK

alexanderwiederin reviewed (commented):
The commit message is inaccurate. `btck_TxValidationResult` does not provide granular failure reasons for this function, only `UNSET` and `CONSENSUS` are reachable via `btck_transaction_check`.

alexanderwiederin commented:
@w0xlt, are you still on this? Apologies if my PR caused the need for rebase.

alexanderwiederin commented on src/kernel/bitcoinkernel.h:
```diff
@@ -344,6 +347,25 @@ typedef uint32_t btck_BlockValidationResult;
+#define btck_TxValidationResult_MEMPOOL_POLICY      ((btck_TxValidationResult)(9))
+#define btck_TxValidationResult_NO_MEMPOOL          ((btck_TxValidationResult)(10))
+#define btck_TxValidationResult_RECONSIDERABLE      ((btck_TxValidationResult)(11))
+#define btck_TxValidationResult_UNKNOWN             ((btck_TxValidationResult)(12))
```
Was this addressed? It looks unchanged to me.

alexanderwiederin replied:
pinging @w0xlt.

w0xlt commented:
@alexanderwiederin  No worries at all! Just rebased, there was only a small conflict in `bitcoinkernel_wrapper.h` which was straightforward to resolve.

sedited commented on src/kernel/bitcoinkernel.h:
```diff
@@ -641,6 +702,30 @@ BITCOINKERNEL_API void btck_precomputed_transaction_data_destroy(btck_Precompute
 
 ///@}
 
+/** @name Consensus
```
We've previously put these "context free" in the section of their respective data structure. That would also make more sense to me, since we generally group by data structure, not type of operation.

w0xlt replied:
Done. Thanks.

sedited reviewed (commented):
ACK 843133c778172134ff4284d341497ba220f7123e

alexanderwiederin commented:
ACK https://github.com/bitcoin/bitcoin/pull/33796/changes/843133c778172134ff4284d341497ba220f7123e

sedited commented:
@janb84 Can you take another look here?

janb84 commented on src/kernel/bitcoinkernel.h:
```diff
@@ -641,6 +702,30 @@ BITCOINKERNEL_API void btck_precomputed_transaction_data_destroy(btck_Precompute
+ *
+ * Performs basic structural consensus checks (consensus/tx_check::CheckTransaction)
+ * without requiring blockchain state.
+ *
```
NIT: maybe lean a bit more into clarifying that it's context-free checking.

```suggestion
 * Does NOT verify scripts/signatures, input availability, locktime, sequence locks, BIP30/BIP68, fees, or mempool policy.
 *
```

w0xlt replied:
Done. Thanks

janb84 commented on src/kernel/bitcoinkernel.h:
```diff
@@ -641,6 +702,30 @@ BITCOINKERNEL_API void btck_precomputed_transaction_data_destroy(btck_Precompute
 ///@}
 
+/** @name Consensus
+ *  Context‑free consensus checks.
```
```suggestion
 *  Context-free consensus checks.
```
Is this non-breaking hyphen (U+2011) intentional ? 

NIT: Replace non ascii hyphen with ascii dash.

w0xlt replied:
The `Consensus` section was dropped (per @sedited grouping suggestion), so `btck_transaction_check` now lives in the `Transaction` group.

janb84 commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -981,6 +997,30 @@ inline bool Block::Check(const ConsensusParamsView& consensus_params,
+{
+    TxValidationState state;
+    const bool ok = btck_transaction_check(tx.get(), state.get()) == 1;
+    return {ok, std::move(state)};
```
This breaks with the style / shape used in `Block::Check`. Are you not now returning the same fact twice ?

w0xlt replied:
Done. Wrapper reshaped to `bool CheckTransaction(const Transaction& tx, TxValidationState& state)`, matching `Block::Check` (caller-owned state, single bool return, no redundant info).

janb84 reviewed (commented):
Concept ACK 843133c778172134ff4284d341497ba220f7123e

Found one non-breaking hyphen (U+2011), is this intentional?  

it's clear enough for a non-core-dev that this function is context-free (and what that implies). Especially the latter, do we communicate clearly (enough) that it will not fully validate the transaction. (so therefore the nit, but it's non blocking !  I'm on the fence myself)

sedited reviewed (approved):
Re-ACK bfbf1a7ef3ee64a345dc2b2c6cf4f85c3a739fae

janb84 reviewed (commented):
ACK bfbf1a7ef3ee64a345dc2b2c6cf4f85c3a739fae

Thanks for incorporating my nits :)

alexanderwiederin commented:
Re-ACK https://github.com/bitcoin/bitcoin/pull/33796/changes/bfbf1a7ef3ee64a345dc2b2c6cf4f85c3a739fae
